lib/web/apiserver.go

```diff
@@ -588,10 +588,29 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 				// legacy app access needs to make a CORS fetch request,
 				// so we only set the default CSP on that page
 				parts := strings.Split(r.URL.Path, "/")
+
+				if len(parts[3]) == 0 {
+					h.log.Warn("Application URL is missing from web/launch URL")
+					http.Error(w, "missing application URL", http.StatusBadRequest)
+					return
+				}
+
 				// grab the FQDN from the URL to allow in the connect-src CSP
-				applicationURL := "https://" + parts[3] + ":*"
+				applicationURL := "https://" + parts[3]
+
+				// Parse to validate the URL extracted is in a valid format.
+				// An invalid URL is:
+				//  - having spaces
+				//  - having escape characters eg: %20
```

Comment:
None of these two are guaranteed by `url.Parse`; they only guarantee that these characters don't appear in the domain name. What am I misssing?

Reply:
> they only guarantee that these characters don't appear in the domain name

hrm, i wasn't exactly sure what you meant, but this is what we want? the parsing will fail if spaces or escape characters like %20 is present?

Reply:
OK, so in this case, the comment is misleading, because it suggests that any URL with spaces is considered invalid; however, an URL that has a space somewhere in the path is valid.

I tried using such an URL, but I guess because %2F is interpreted verbatim on the path resolution stage, the `applicationURL` variable is guaranteed to have just the domain. This would mean we're fine here, but the comment is still misleading. Otherwise, LGTM.

Reply:
ah gotcha, thanks for catching that. i changed the comments/naming to use correct terms: https://github.com/gravitational/teleport/pull/41236/commits/218801c33a96bec943fa903a1d57ab581ce47088